src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,612 @@
+      }
+      int local_s_id = 0; // There is always a root on the schema
+      int local_c_id = -1; // for columns it is just the leaf nodes
+      auto num = names.size();
+      std::vector<column_pruner*> tree_stack;
+      std::vector<int> num_children_stack;
+      tree_stack.push_back(this);
+      num_children_stack.push_back(parent_num_children);
+      for(uint64_t i = 0; i < num; i++) {
+        auto name = names[i];
+        auto num_c = num_children[i];
+        local_s_id++;
+        int tmp_c_id = -1;
+        if (num_c == 0) {
+          // leaf node...
+          local_c_id++;
+          tmp_c_id = local_c_id;
+        }
+        tree_stack.back()->children.try_emplace(name, local_s_id, tmp_c_id);
+        if (num_c > 0) {
+          tree_stack.push_back(&tree_stack.back()->children[name]);
+          num_children_stack.push_back(num_c);
+        } else {
+          // go back up the stack/tree removing children until we hit one with more children
+          bool done = false;
+          while (!done) {
+              int parent_children_left = num_children_stack.back() - 1;
+              if (parent_children_left > 0) {
+                num_children_stack.back() = parent_children_left;
+                done = true;
+              } else {
+                tree_stack.pop_back();
+                num_children_stack.pop_back();
+              }
+
+              if (tree_stack.size() <= 0) {
+                done = true;
+              }
+          }
+        }
+      }
+      if (tree_stack.size() != 0 || num_children_stack.size() != 0) {
+        throw std::invalid_argument("DIDN'T CONSUME EVERYTHING...");
+      }
+    }
+
+    std::map<std::string, column_pruner> children;
+    // The following IDs are the position that they should be in when output in a filteres footer, except
+    // that if there are any missing columns in the actual data the gaps need to be removed.
+    // schema ID
+    int s_id;
+    // Column chunk and Column order ID
+    int c_id;
+};
+
+static bool invalid_file_offset(long start_index, long pre_start_index, long pre_compressed_size) {
+  bool invalid = false;
+  // checking the first rowGroup
+  if (pre_start_index == 0 && start_index != 4) {
+    invalid = true;
+    return invalid;
+  }
+
+  //calculate start index for other blocks
+  int64_t min_start_index = pre_start_index + pre_compressed_size;
+  if (start_index < min_start_index) {
+    // a bad offset detected, try first column's offset
+    // can not use minStartIndex in case of padding
+    invalid = true;
+  }
+
+  return invalid;
+}
+
+static int64_t get_offset(parquet::format::ColumnChunk & column_chunk) {
+  auto md = column_chunk.meta_data;
+  int64_t offset = md.data_page_offset;
+  if (md.__isset.dictionary_page_offset && offset > md.dictionary_page_offset) {
+    offset = md.dictionary_page_offset;
+  }
+  return offset;
+}
+
+static std::vector<parquet::format::RowGroup> filter_groups(parquet::format::FileMetaData & meta, 
```

Comment:
```suggestion
static std::vector<parquet::format::RowGroup> filter_groups(parquet::format::FileMetaData const& meta, 
```

Reply:
though I am not 100% sure if that will compile with all the requirements here. It would just be nice.